sets/training_sets.py

```diff
@@ -0,0 +1,336 @@
+# -*- coding: utf-8 -*-
+"""Generate TrainingSets Sets"""
+
+import abc
+import os
+from functools import reduce
+
+import exifread
+import lxml.etree as etree
+import numpy as np
+
+from cv2 import cv2
+from PIL import (
+    Image
+)
+from shapely.geometry import (
+    Polygon
+)
+
+
+XML_NS = {
+    'alto': 'http://www.loc.gov/standards/alto/ns-v3#',
+    'page2013': 'http://schema.primaresearch.org/PAGE/gts/pagecontent/2013-07-15',
+    'page2019': 'http://schema.primaresearch.org/PAGE/gts/pagecontent/2019-07-15'}
+
+DEFAULT_MIN_CHARS = 16
+
+DEFAULT_DPI = 300
+
+
+class TextLine(abc.ABC):
+    """
+    TextLine from structured OCR-Data
+    """
+
+    def __init__(self, element, namespace, revert=False):
+        self.element = element
+        self.namespace = namespace
+        self.element_id = None
+        self.set_id()
+        self.text_words = []
+        self.set_text_words()
+        self.revert = revert
+        self.box = self.to_box(self.element)
+
+    @abc.abstractmethod
+    def set_id(self):
+        """Determine identifier"""
+
+    @abc.abstractmethod
+    def set_text_words(self):
+        """Determine list of word tokens"""
+
+    @abc.abstractmethod
+    def to_box(self, element):
+        """Return bounding box of TexLine shape"""
+
+    def get_textline_content(self) -> str:
+        """
+        Set TextLine contents from it's included word tokens
+        may revert order of tokens if required
+        """
+
+        if self.revert:
+            return reduce(lambda c, p: p + ' ' + c, self.text_words)
```

Comment:
Since we already require python-bidi, it would probably more robust to use it for handling the inversion, c.f. https://github.com/MeirKriheli/python-bidi/blob/master/bidi/algorithm.py / https://github.com/MeirKriheli/python-bidi#api

Reply:
Thanks for the hint, I'll take a look!

Reply:
Hm, 
I guess we need to go without bidi so far, since it looks the output from mixed arab + latin lines makes them turn from rtl to ltr. Lines with only arab chars and indic numbers seem to work pretty well with bidi, but mixed don't.

I'm no export on that, though. Please, take a look yourself. I've added the bidi import, adopted line content generation (the commented section). Feel free to switch implementations. (in your preferred IDE place a debugging mark in `test_create_sets_from_page2013_and_jpg` to inspect the temporary test files written)

Reply:
Do I get it right: bidi works on char-Level? If so, I don't think it is useful in this scenario. I only know some (rather poor) arabic output generated from tesseract itself which is word-based.